tonic/functional/event_downsampling.py

```diff
@@ -0,0 +1,271 @@
+    """Feedback inhibition to eliminate high-frequency noise inspired by locust brains.
+    Parameters:
+        events (ndarray): ndarray of shape [num_events, num_event_channels].
+        target_size (tuple): a 2-tuple of x,y denoting new down-sampled size for events to be
+                             re-scaled to (new_width, new_height).
+        dt (float): step size for simulation, in ms.
+        buffer_layer (dict): dictionary of neuron population parameters.
+        inhibitory_layer (dict): dictionary of neuron population parameters.
+        buffer_postsynaptic_strength (float): synaptic strength of buffer to inhibitory neuron populations.
+        inhibitory_postsynaptic_strength (float): synaptic strength of inhibitory to buffer neuron populations.
+        delay (int): dendritic delay in timesteps.
+    Returns:
+        the downsampled events after feedback inhibition.
+        
+    * Define leaky integrate and fire (LIF) and linear threshold (LT) neuron models
+    V(t+1) = p*V(t) + I_syn
+    V (ndarray): membrane potential of shape target_size, initial value of V set at zero
+    p (float): persistence variable or leak factor is a float that is an element of {0, 1}
+    I_syn (nd_array): injection current to layer of neurons of shape target_size
+    
+    where
+    I_syn = g_exc * w_e * a_e(t-delta_e) - g_inh * w_i * a_i(t-delta_i)
+    g_{exc, inh} (float): synaptic strength that is an element of {0, 1}
+    a_{e, i} (ndarray): excitatory neuron population activity
+    delta_{e, i} (int): timestep delays
+    
+    * Activity updates
+    LT model -> a=v(t+1) if v(t+1)>=theta, else a=0
+    LIF model -> a=1 if v(t+1)>=theta, else a=0
+    After hyperpolarisation, v'(t+1) = v(t+1) - alpha where alpha is the amplitude of hyperpolarisation
+    
+    Assumptions: One-to-one synaptic connectivities
+                 Homogeneous layer-wise neuron and synapse population attributes
+    """
+    
+    assert "x" and "y" in events.dtype.names
+    
+    events = events.copy()
+    
+    events = time_bin_numpy(events, dt)
+    
+    # Parameter definitions
+    assert "p" and "theta" and "alpha" in buffer_layer_parameters.keys()
+    assert "p" in inhibitory_layer_parameters.keys()
+    
+    # All event times
+    event_times = np.unique(events["t"])
+    
+    # Separate by polarity
+    events_positive = events[events["p"] == 1]
+    events_negative = events[events["p"] == 0]
+    
+    assert isinstance(delay, int)
+    
+    simulation_time = event_times[-1] + (2 + delay) * dt + dt
+    simulation_timesteps = int(np.ceil(simulation_time / dt))
+    
+    # Ravelled array of neuron population
+    buffer_layer = np.zeros((simulation_timesteps, np.prod(target_size), 2))
+    inhibitory_layer = np.zeros((simulation_timesteps, np.prod(target_size), 2))
+    
+    for t, time in enumerate(np.arange(0, event_times[-1] + dt, dt)):
```

Comment:
is there a way to vectorize that? iterating in a for loop is very slow for recordings with millions of events

Reply:
Removed this method.